highfidelity/hifi
# Update Oculus integration

Addresses a number of bugs / annoyances, including [this one](https://highfidelity.fogbugz.com/f/cases/14164/Exit-VR-in-world-not-working-properly-on-Oculus) to which the others are childed.  

This PR

* Updates our Oculus integration to the latest SDK (1.26)
* Enables Oculus Dash integration
  * Pressing the Oculus button on the right touch controller will now bring up oculus dash 'in-world'
  * When in Oculus dash mode, we stop updating the Avatar hand poses because the Oculus dash renders it's own hands
* Enables a better mixed mode experience
  * When switching out of VR, we stop submitting frames to the HMD, but we also disconnect the rendering session in such a way that the Oculus runtime will take over and start rendering the dash again.  This prevents the freezes of the last seen frame when exiting VR mode via a variety of ways.  

## Developer notes

The improved functionality is implemented by having the Oculus display plugin have a dedicated rendering session while it's active.  When the Oculus display plugin is _inactive_ we destroy the session and de-initialize the Oculus SDK, and then re-initialize it with a non-rendering session, using the Oculus flag `ovrInit_Invisible`, which tells the Oculus runtime that we're not intending to do any rendering to the HMD, but still allows us to query the session state (like whether the HMD is currently being worn or not) and the hand controllers.

Dash integration simply requires passing the `ovrInit_FocusAware` flag to the Oculus initialization function.  Implementation of the dash _could_ be improved if in addition to the scene color texture, we also provided Oculus with a scene depth texture, but this would require significant changes making it out of scope for this PR.  

## Testing

Testing requires an Oculus HMD and touch controllers.  
*Make sure* you have "Rift Core 2.0 Beta" enabled in the Settings -> Beta section of the Oculus desktop application.

* Start interface.  
* Switch to VR mode and put on the HMD
* Verify hand controllers work as expected
* Press the Oculus button on the right hand controller
  * In the master build, the Interface VR environment rendered by interface will stop rendering and you'll be in an empty space with the Oculus menu
  * In the PR build, the avatar hands should revert to their default pose, and you should see the Oculus dash super-imposed over the Interface VR environment
* Switch out of the HMD either using the keyboard shortcut for desktop mode (Ctrl+1) or using the tablet and selecting "Exit VR"
  * In the master build, the last frame rendered in VR mode will simply freeze on the HMD
  * In the PR build, the HMD should switch to rendering Oculus content
* Exit interface

This PR refactors the Oculus plugin quite a bit, so we should run regression testing for any Oculus or HMD specific functionality, particularly around entering and exiting VR, donning or removing the HMD, and using the hand controllers.
